calyx/opt/src/passes/constant_port_prop.rs

```diff
@@ -0,0 +1,149 @@
+    }
+
+    fn description() -> &'static str {
+        "propagates constants when port values can be inferred"
+    }
+}
+
+impl Visitor for ConstantPortProp {
+    fn invoke(
+        &mut self,
+        s: &mut calyx_ir::Invoke,
+        _comp: &mut calyx_ir::Component,
+        _sigs: &LibrarySignatures,
+        _comps: &[calyx_ir::Component],
+    ) -> VisResult {
+        if let Some(cg) = &s.comb_group {
+            let cg_name = cg.borrow().name();
+            let cell_name = s.comp.borrow().name();
+            if let Some(registered_cell_name) =
+                self.comb_groups_to_modify.get(&cg_name)
+            {
+                if *registered_cell_name != cell_name {
+                    // there is a different invoke that is using the same comb group
+                    self.comb_groups_used_elsewhere.insert(cg_name);
+                }
+            } else {
+                // no invokes have used this comb group so far
+                self.comb_groups_to_modify.insert(cg_name, cell_name);
+            }
+        }
+        Ok(Action::Continue)
+    }
+
+    fn start_while(
+        &mut self,
+        s: &mut calyx_ir::While,
+        _comp: &mut calyx_ir::Component,
+        _sigs: &LibrarySignatures,
+        _comps: &[calyx_ir::Component],
+    ) -> VisResult {
+        if let Some(comb_group) = &s.cond {
+            self.comb_groups_used_elsewhere
+                .insert(comb_group.borrow().name());
+        }
+        Ok(Action::Continue)
+    }
+
+    fn start_if(
+        &mut self,
+        s: &mut calyx_ir::If,
+        _comp: &mut calyx_ir::Component,
+        _sigs: &LibrarySignatures,
+        _comps: &[calyx_ir::Component],
+    ) -> VisResult {
+        if let Some(comb_group) = &s.cond {
+            self.comb_groups_used_elsewhere
+                .insert(comb_group.borrow().name());
+        }
+        Ok(Action::Continue)
+    }
+
+    fn finish(
+        &mut self,
+        comp: &mut calyx_ir::Component,
+        sigs: &LibrarySignatures,
+        _comps: &[calyx_ir::Component],
+    ) -> VisResult {
+        let mut builder = ir::Builder::new(comp, sigs);
+        let one = builder.add_constant(1, 1);
+        let zero = builder.add_constant(0, 1);
+        // first, drop any comb groups that are used in while/ifs and in multiple invokes
+        for used_comb_group in &self.comb_groups_used_elsewhere {
+            self.comb_groups_to_modify.remove(used_comb_group);
+        }
+        // modify assignments of any remaining comb groups
+        for comb_group_ref in comp.comb_groups.iter() {
+            let mut comb_group = comb_group_ref.borrow_mut();
+            let comb_group_name = comb_group.name();
+            if let Some(cell_name) =
+                self.comb_groups_to_modify.get(&comb_group_name)
+            {
+                let mut modified_asgns =
+                    std::mem::take(&mut comb_group.assignments);
+                for asgn in &mut modified_asgns {
```

Comment:
This is definitely a personal preference thing, but I think I prefer
`.iter_mut()` here over using the `IntoIterator` trait